src/axolotl/utils/callbacks.py

```diff
@@ -317,3 +320,250 @@ def on_evaluate(
                 trainer.log(results)
 
     return BenchEvalCallback
+
+
+def log_prediction_callback_factory(trainer: Trainer, tokenizer):
+    LOG.info("log_prediction_callback_factory")
+
+    class LogPredictionCallback(TrainerCallback):
+        """Callback to log prediction values during each evaluation"""
+
+        def __init__(self, cfg):
+            self.cfg = cfg
+            self.logged = False
+
+        def on_evaluate(
+            self,
+            args: AxolotlTrainingArguments,
+            state: TrainerState,
+            control: TrainerControl,
+            model,
+            # tokenizer,
+            train_dataloader,
+            eval_dataloader,
+            **kwargs,
+        ):
+            LOG.info("=" * 80)
+            LOG.info("logging predictions")
+
+            trainer.model.eval()
+
+            def logits_to_tokens(logits) -> str:
+                probabilities = torch.softmax(logits, dim=-1)
+                # Get the predicted token ids (the ones with the highest probability)
+                predicted_token_ids = torch.argmax(probabilities, dim=-1)
+                return predicted_token_ids
+
+            def log_table_from_dataloader(name: str, table_dataloader):
+
+                # Initialize an empty wandb.Table
+                table = wandb.Table(columns=["id", "Prompt", "Correct Completion", "Predicted Completion 1", "Predicted Completion 2"])
+
+                # preds, refs = [], []
+                # loss_bench = 0
+                # predictions = []
+                id = 0
+                for batch in tqdm(table_dataloader, total=len(table_dataloader)):
+                # max_examples = 100
+                # for batch in tqdm(table_dataloader, total=min(max_examples, len(table_dataloader))):
+
+                    # batch.data['labels'].shape
+                    # torch.Size([2, 320])
+                    # values at front with -100 are supposed to be prompt tokens
+                    # values after are completion tokens
+
+                    # batch.data['input_ids'].shape
+                    # torch.Size([2, 320])
+                    
+                    # # Extract prompt and completion tokens from input_ids based on labels
+                    # prompt_token_ids = batch.data['input_ids'][batch.data['labels'] == IGNORE_INDEX]
+                    # completion_token_ids = batch.data['input_ids'][batch.data['labels'] != IGNORE_INDEX]
+
+                    # # prompt_texts = tokenizer.batch_decode(batch.data['input_ids'])
+                    # prompt_texts = tokenizer.batch_decode(prompt_token_ids)
+                    # completion_texts = tokenizer.batch_decode(completion_token_ids)
+
+                    (loss, logits, labels) = trainer.prediction_step(
+                        trainer.model,
```

Comment:
**Help Wanted**
Is this the best way to recompute the model's prediction? I don't think the `on_evaluate` callback has access to the actual prediction outputs so I need to recompute them, right?

See https://github.com/huggingface/transformers/blob/0afa5071bd84e44301750fdc594e33db102cf374/src/transformers/trainer.py#L2968-L2996

Reply:
you should have access to `model` , right? if so, I think `model(inputs)` would work?

Reply:
I originally had both `trainer.prediction_step` and `trainer.model.generate`. Now I've removed `prediction_step` since the output was also not useful and leaving `model.generate` as you recommended here as well.